workflow/controller/taskresult.go

```diff
@@ -53,21 +62,100 @@ func (wfc *WorkflowController) newWorkflowTaskResultInformer() cache.SharedIndex
 	return informer
 }
 
-func (woc *wfOperationCtx) taskResultReconciliation() {
+func shouldErrorPodTimeout(node *wfv1.NodeStatus) bool {
+	return time.Since(node.StartedAt.Time) <= envutil.LookupEnvDurationOr("POD_ABSENT_TIMEOUT", 2*time.Minute)
+}
+
+func (woc *wfOperationCtx) taskResultReconciliation() error {
+
 	objs, _ := woc.controller.taskResultInformer.GetIndexer().ByIndex(indexes.WorkflowIndex, woc.wf.Namespace+"/"+woc.wf.Name)
 	woc.log.WithField("numObjs", len(objs)).Info("Task-result reconciliation")
+
+	taskresultCli := woc.controller.wfclientset.ArgoprojV1alpha1().WorkflowTaskResults(woc.wf.Namespace)
+	// We will be hitting the cache here, should we bypass the cache?
+	podList, err := woc.getAllWorkflowPods()
+	if err != nil {
+		woc.log.Warnf("couldn't retrieve pods")
+		return err
+	}
+	podMap := make(map[string]*apiv1.Pod)
+
+	// can use woc.nodePodExit technically
+	for _, pod := range podList {
+		pod := pod
+		nodeID := woc.nodeID(pod)
+		podMap[nodeID] = pod
+	}
+
 	for _, obj := range objs {
 		result := obj.(*wfv1.WorkflowTaskResult)
 		resultName := result.GetName()
 
 		woc.log.Debugf("task result:\n%+v", result)
 		woc.log.Debugf("task result name:\n%+v", resultName)
 
+		label, foundLabel := result.Labels[common.LabelKeyReportOutputsCompleted]
+
+		pod, foundPod := podMap[result.Name]
+		node, err := woc.wf.Status.Nodes.Get(result.Name)
+		if err != nil {
+			if foundPod {
+				// how does this path make any sense?
+				// pod created but informed not yet updated
+				woc.log.Errorf("couldn't obtain node for %s, but found pod, this is not expected, doing nothing", result.Name)
+			}
+			continue
+		}
+
+		if foundPod {
+			woc.log.Debugf("Got pod %s with phase %s for task result %s and node id %s with label %s", pod.Name, pod.Status.Phase, resultName, result.Name, label)
+			woc.log.Debugf("The node phase was %s for node named %s", node.Phase, node.Name)
+		} else if !foundPod && !node.Completed() {
+			woc.log.Debugf("couldn't find pod")
+			timeout := shouldErrorPodTimeout(node)
+			if timeout {
+				woc.log.Infof("Determined controller should timeout for %s", result.Name)
+				result.Labels[common.LabelKeyReportOutputsCompleted] = "true"
+
+				// patch the label
```

Comment:
I just thought it was nicer if the wfTaskResult was also updated the moment we decided to treat this wfTaskResult as completed.

Currently this would still work, but the completed value would be false. This is fine because it would get cleaned up by the controller anyway.

We can ensure we hit the api exactly once though, by checking if the label doesn't exist or isnt' true before patching it.

Reply:
This is still an expensive operation to perform, and for no-one's benefit that I can see.

This workflow will stop caring about this WorkflowTaskResult after this moment, and no-one else does. I'd rather it was deleted than patched. I'm pretty sure it'll get deleted fairly shortly after this anyway.